// wikifetch/lib/wikifetch.js
var jQuery = require('jquery'),
		request = require('request'),
		sexy = require('sexy-args');

function WikiFetch(params) {
	sexy.args([this, 'object1'], {
		'object1': {
			articlePrefix: 'http://en.wikipedia.org/wiki/'
		}
	}, function() {
		sexy.extend(this, params);
	});
};

WikiFetch.prototype.loadArticle = function(articleName, callback) {
	//console.log('loadArticle'.green);
	//console.log(this.articlePrefix + articleName);
	request({
		url: this.articlePrefix + articleName,
		headers: {
    'User-Agent': 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.8.1.13) Gecko/20080311 Firefox/2.0.0.13'
  	}
	}, function(err, res, body) {
		console.log('req response'.zebra);
		var error = err || (res.statusCode != 200 ? res.statusCode : false);
		if (error) {
			callback(error, null);
			return;
		}

		var wiki_result = {};

		// TITLE
		wiki_result.title = jQuery(body).find('#firstHeading').text();

		// SUMMARY
		var b = jQuery(body).find('#mw-content-text > p');
		var first = false;
		b.each(function() { // OTTO HALPS ME!
			if(!first){
				var element = jQuery(this);
				if(element.text().match(/may refer to:/g)){
					wiki_result.topics = [];

					// Display options
					jQuery(body).find('#mw-content-text h2, #mw-content-text ul li a').not('.toc li, #toctitle h2').each(function(){
						var el = jQuery(this);
						if(el.is('a')){
							//console.log(wiki_result.topics.length);
							if(wiki_result.topics.length > 0){
								wiki_result.topics[wiki_result.topics.length-1].items.push(el.text());
								wiki_result.topics[wiki_result.topics.length-1].links.push(el.attr('href').replace(/^\//, 'http://en.wikipedia.org/'));
								console.log('===>', el.attr('href'));
								console.log('-', el.text());
							}
						}else{
							var disambig = {};
							disambig.title = el.text();
							disambig.items = [];
							disambig.links = [];
							wiki_result.topics.push(disambig);
							console.log('h2', el.text());
						}
					});


				}

				// Remove citation numbers
				wiki_result.summary = element.text().replace(/\[[0-9]+\]/g, '');

				console.log(wiki_result.topics);
				first = true;
			}
		});

		// IMAGE
		var g = jQuery(body).find('#mw-content-text img');
		first = false;
		g.each(function(){
			var element = jQuery(this);
			if(element.is('img') && element.attr('width') > 50 && !first){
				console.log(element.attr('src').replace('//', 'http://'));
				wiki_result.img_src = element.attr('src').replace('//', 'http://');
				first = true;
			}
		});

		callback(null, wiki_result);
	});
};

exports.WikiFetch = WikiFetch;
